src/Analyzers/CSharp/Analyzers/MakeStructReadOnly/CSharpMakeStructReadOnlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,112 @@
+        => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+    protected override void InitializeWorker(AnalysisContext context)
+        => context.RegisterSymbolStartAction(context =>
+        {
+            var compilation = context.Compilation;
+            if (compilation.LanguageVersion() < LanguageVersion.CSharp7_2)
+                return;
+
+            // First, see if this at least strongly looks like a struct that could be converted.
+            if (!IsCandidate(context, out var typeDeclaration, out var option))
+                return;
+
+            // Looks good.  However, we have to make sure that the struct has no code which actually overwrites 'this'
+
+            var writesToThis = false;
+            context.RegisterSyntaxNodeAction(context =>
+            {
+                var semanticModel = context.SemanticModel;
+                var thisExpression = (ThisExpressionSyntax)context.Node;
+                var cancellationToken = context.CancellationToken;
+                writesToThis = writesToThis || thisExpression.IsWrittenTo(semanticModel, cancellationToken);
+            }, SyntaxKind.ThisExpression);
+
+            context.RegisterSymbolEndAction(context =>
+            {
+                // if we wrote to 'this', then we cannot convert this struct.
+                if (writesToThis)
+                    return;
+
+                context.ReportDiagnostic(DiagnosticHelper.Create(
+                    Descriptor,
+                    typeDeclaration.Identifier.GetLocation(),
+                    option.Notification.Severity,
+                    additionalLocations: ImmutableArray.Create(typeDeclaration.GetLocation()),
+                    properties: null));
+            });
+        }, SymbolKind.NamedType);//.RegisterSyntaxNodeAction(AnalyzeSyntax, SyntaxKind.StructDeclaration, SyntaxKind.RecordStructDeclaration);
+
+    private static bool IsCandidate(
+        SymbolStartAnalysisContext context,
+        [NotNullWhen(true)] out TypeDeclarationSyntax? typeDeclaration,
+        [NotNullWhen(true)] out CodeStyleOption2<bool>? option)
+    {
+        var typeSymbol = (INamedTypeSymbol)context.Symbol;
+        var cancellationToken = context.CancellationToken;
+
+        typeDeclaration = null;
+        option = null;
+        if (typeSymbol.TypeKind is not TypeKind.Struct)
+            return false;
+
+        if (typeSymbol.IsReadOnly)
+            return false;
+
+        if (typeSymbol.DeclaringSyntaxReferences.Length == 0)
+            return false;
+
+        typeDeclaration = typeSymbol.DeclaringSyntaxReferences[0].GetSyntax(cancellationToken) as TypeDeclarationSyntax;
+        if (typeDeclaration is null)
+            return false;
+
+        var options = new CSharpAnalyzerOptionsProvider(context.Options.AnalyzerConfigOptionsProvider.GetOptions(typeDeclaration.SyntaxTree), context.Options);
+        option = options.PreferReadOnlyStruct;
```

Comment:
I think other analyzers use a more simple way to get the option.

Reply:
this is the first symbol-analzyer that needs it.

Reply:
Consider adding an extension method here:

https://github.com/dotnet/roslyn/blob/d73d31cbccb9aa850f3582afb464b709fef88fd7/src/Analyzers/CSharp/Analyzers/CodeStyle/CSharpAnalyzerOptionsProvider.cs#L125

```csharp
    public static CSharpAnalyzerOptionsProvider GetCSharpAnalyzerOptions(this SymbolAnalysisContext context)
        => new(context.Options.AnalyzerConfigOptionsProvider.GetOptions(context.Symbol.Locations[0].SourceTree), context.Options);
```

Or if it's not safe to access `Locations[0]`, consider adding a `SyntaxTree` parameter.